src/analyses/base.ml

```diff
@@ -2402,6 +2530,54 @@ struct
         | Some x -> assign ctx x (List.hd args)
         | None -> ctx.local
       end
+    (**Floating point classification and trigonometric functions defined in c99*)
+    | Math { args; }, _ ->
+      begin match args with
+        | [x] -> 
+          let eval_x = eval_rv (Analyses.ask_of_ctx ctx) gs st x in
+          begin match eval_x with
+            | `Float float_x -> 
+              let result = 
+                begin match f.vname with
+                  | "__builtin_isfinite" -> `Int (ID.cast_to IInt (FD.isfinite float_x))
+                  | "__builtin_isinf" | "__builtin_isinf_sign" -> `Int (ID.cast_to IInt (FD.isinf float_x))
+                  | "__builtin_isnan" -> `Int (ID.cast_to IInt (FD.isnan float_x))
+                  | "__builtin_isnormal" -> `Int (ID.cast_to IInt (FD.isnormal float_x))
+                  | "__builtin_signbit" -> `Int (ID.cast_to IInt (FD.signbit float_x))
+                  | "__builtin_acos" | "acos" -> `Float (FD.acos float_x)
+                  | "__builtin_asin" | "asin" -> `Float (FD.asin float_x)
+                  | "__builtin_atan" | "atan" -> `Float (FD.atan float_x)
+                  | "__builtin_cos" | "cos" -> `Float (FD.cos float_x)
+                  | "__builtin_sin" | "sin" -> `Float (FD.sin float_x)
+                  | "__builtin_tan" | "tan" -> `Float (FD.tan float_x)
+                  | _ -> failwith (f.vname^" should be implemented in goblint but isn't")
```

Comment:
There's somewhat of a duplication of all these names between `LibraryFunctions` and here. Also the fact that the multiple versions of the same function (without and with `__builtin_`) need to be handled everywhere, which defeats the point of using `LibraryFunctions`.

A nicer approach might be to define an auxiliary type like
```ocaml
type math =
  | Sin of exp
  (* ... *)
  | Atan2 of exp * exp
```
and have the outer classification constructor as 
```ocaml
  | Math of math
```

Not only does that avoid the name duplication, it also avoids all these `failwith` cases that are currently here, cluttering the interesting parts. With proper `LibraryDesc`s, the checking of argument counts would all be done by `LibraryFunctions` with its existing functionality, like it is for `Memset` above.

Reply:
Good Idea. It is implemented now in this way.